stock_putaway_storage_type_strategy/models/stock_location.py

```diff
@@ -0,0 +1,160 @@
+        quant = self.env.context.get('putaway_quant')
+        package_storage_type = False
+        if quant:
+            package_storage_type = quant.package_id.packaging_id.stock_package_storage_type_id
+        if not package_storage_type:
+            return super()._get_putaway_rule_location(
+                product=product, category=category
+            )
+        putaway_rules = self.putaway_rule_ids.filter_rules(
+            product=product, category=category
+        )
+        if category and not product:
+            product = quant.product_id
+        for rule in putaway_rules:
+            dest_location = rule._get_destination_location()
+            if (
+                dest_location.pack_storage_strategy == 'none' and
+                dest_location._package_storage_type_allowed(
+                    package_storage_type
+                )
+            ):
+                return dest_location
+            storage_locations = dest_location.get_storage_locations(
+                product=product
+            )
+            allowed_location = storage_locations.select_allowed_location(
+                package_storage_type, product=product
+            )
+            if allowed_location:
+                return allowed_location
+        return self.browse()
+
+    def get_storage_locations(self, product=None):
+        self.ensure_one()
+        locations = self.browse()
+        if self.pack_storage_strategy == 'none':
+            locations = self
+        elif self.pack_storage_strategy == 'ordered_locations':
+            locations = self._get_ordered_children_locations()
+        return locations
+
+    def select_allowed_location(self, package_storage_type, product=None):
+        for location in self:
+            if (
+                location._package_storage_type_allowed(
+                    package_storage_type, product=product
+                )
+            ):
+                return location
+        return self.browse()
+
+    def _get_ordered_children_locations(self):
+        return self.search(
+            [('id', 'child_of', self.ids), ('id', '!=', self.id)]
+        )
+
+    def _package_storage_type_allowed(self, package_storage_type, product=None):
+        self.ensure_one()
+        matching_location_storage_types = self.allowed_stock_location_storage_type_ids.filtered(
+            lambda slst: package_storage_type in slst.stock_package_storage_type_ids
+        )
+        allowed_location_storage_types = self.filter_restrictions(
+            matching_location_storage_types, product=product
+        )
+        return not self.allowed_stock_location_storage_type_ids or allowed_location_storage_types
+
+    def filter_restrictions(self, matching_location_storage_types, product=None):
+        allowed_location_storage_types = self.env['stock.location.storage.type']
+        for location_storage_type in matching_location_storage_types:
+            if location_storage_type.only_empty:
```

Comment:
Should these 'strategies' be called using built functions with getattr('xxx_only_empty') to ease override ?